abi-bindings/go/packer/packer_test.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package packer
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"math/big"
+	"math/rand"
+	"reflect"
+	"testing"
+
+	validatorsetsig "github.com/ava-labs/teleporter/abi-bindings/go/governance/ValidatorSetSig"
+	teleportermessenger "github.com/ava-labs/teleporter/abi-bindings/go/teleporter/TeleporterMessenger"
+	teleporterregistry "github.com/ava-labs/teleporter/abi-bindings/go/teleporter/registry/TeleporterRegistry"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/tools/go/packages"
+)
+
+const interfaceName = "ABIPacker"
+
+// Paths below are relative to the current package.
+const packagePath = "../..."
+const currentPackagePath = "."
+
+var fs = token.NewFileSet()
+
+var packerTypes = map[string]ABIPacker{
+	"ValidatorSetSigMessage": &validatorsetsig.ValidatorSetSigMessage{},
+	"TeleporterMessage":      &teleportermessenger.TeleporterMessage{},
+	"ProtocolRegistryEntry":  &teleporterregistry.ProtocolRegistryEntry{},
+}
+
+func findAllImplementers(t *testing.T) []string {
+	cfg := &packages.Config{
+		Mode: packages.NeedName | packages.NeedFiles | packages.NeedSyntax | packages.NeedTypes | packages.NeedTypesInfo,
+		Fset: fs,
+	}
+	interfacePkg, err := packages.Load(cfg, currentPackagePath)
+	require.NoError(t, err)
+
+	var iface *types.Interface
+	// find the interface in the currnet package
+	for _, pkg := range interfacePkg {
+		for _, obj := range pkg.Types.Scope().Names() {
+			if obj == interfaceName {
+				ifaceObj := pkg.Types.Scope().Lookup(obj)
+				if ifaceObj == nil {
+					continue
+				}
+				if typ, ok := ifaceObj.Type().Underlying().(*types.Interface); ok {
+					iface = typ
+					break
+				}
+			}
+		}
+	}
+	require.NotNilf(t, iface, "Interface %s not found in package %s", interfaceName, currentPackagePath)
+
+	pkgs, err := packages.Load(cfg, packagePath)
+	require.NoError(t, err)
+
+	allImplementers := []string{}
+
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			// Inspect types in the package
+			for _, decl := range file.Decls {
+				if genDecl, ok := decl.(*ast.GenDecl); ok {
+					for _, spec := range genDecl.Specs {
+						if typeSpec, ok := spec.(*ast.TypeSpec); ok {
+							obj := pkg.Types.Scope().Lookup(typeSpec.Name.Name)
+							if obj == nil {
+								continue
+							}
+							if _, ok := obj.Type().Underlying().(*types.Struct); ok {
+								typeName := typeSpec.Name.Name
+								obj := pkg.Types.Scope().Lookup(typeName)
+								if obj != nil {
+									if named, ok := obj.Type().(*types.Named); ok {
+										if types.Implements(types.NewPointer(named), iface) {
+											allImplementers = append(allImplementers, named.Obj().Name())
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
I'm not exactly sure what this part of the test is doing, but generally when something has this many nested statements it gets hard to read. Can the if statements be inverted (i.e. return early in failure case and have no nesting beneath it), and loop bodies broken out into helpers?

Reply:
I absolutely agree in general.   Ast traversal for very particular types is the  only reason why I would do nesting like this without breaking it out into helpers. I prefer it like this for 2 reasons:
:one: The flow here is of zeroing down on a particular thing. In this case instances of the `ABIPacker` interface. With each nesting you are closer to the specific thing that you are looking for.
:two: If breaking out in helpers the interfaces types on the functions wouldn't be helpful. These are all `types.Object` interfaces and get cast to different more specific interfaces (or fail the cast). 

Also FWIW. Github makes it look a lot more egregious visually since it uses 8 spaces for a tab.

It does sound like we have two people who prefer doing something to this so I would be happy to but I just wanted to get my 2 cents in 😄

Reply:
I unrolled them and added some comments. I think I still prefer the previous structure but it sounds like I'm the odd one out 😄